support declaring icds toggles via localsettings

sravfeyn opened the pull request:
Building off of the discussion from https://github.com/dimagi/commcare-hq/pull/26057

I have ran a script to get all enabled/disabled domain toggles on ICDS and declared them as always_enabled/always_disabled here https://github.com/dimagi/commcare-cloud/pull/3435

@snopoke @emord

sravfeyn committed:
supprot declaring icds toggles via localsettings

sravfeyn committed:
fix operator

sravfeyn committed:
spell

snopoke commented on corehq/toggles.py:
```diff
@@ -93,8 +93,8 @@ def index(self):
-                 description=None, save_fn=None, always_enabled=None,
-                 always_disabled=None, enabled_for_new_domains_after=None,
+                 description=None, save_fn=None, always_enabled=set(),
+                 always_disabled=set(), enabled_for_new_domains_after=None,
```
default values should never be mutable: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

sravfeyn replied:
Thanks, missed that.

snopoke commented on corehq/toggles.py:
```diff
@@ -106,8 +106,12 @@ def __init__(self, slug, label, tag, namespaces=None, help_link=None,
+        # For icds-cas domain, toggles are delcared in localsettings statically
+        #   to avoid cache lookups
+        self.always_enabled = (always_enabled |
+            {'icds-cas'} if self.slug in settings.ICDS_CAS_ALWAYS_ENABLED_TOGGLES else {})
```
I'm not a huge fan of having this be so ICDS specific

snopoke reviewed (commented):
I'm not sure there is enough value in having them as settings instead of statically declared in code. I know it allows changing them without a code deploy but given the frequency with which they change that doesn't seem too necessary.

If we could do it in a way that wasn't ICDS specific then I think I'd be ok with it.

I think possibly the best way to do it would be similar to how the Django DB routers work where there is a default router that applies the default logic but can be overridden if required. This would be a bigger code change though but I think it could be done without being too invasive by calling the configurable check class / function in `StaticToggle.enabled`. May also need to provide a way to know which static toggles are enabled for the UI.

Alternatively the setting you have here could be restructured as either:
```
STATIC_TOGGLES_STATE = {
    'domain': {
        'enabled': [....],
        'disabled': [....]
    }
}
```
or
```
STATIC_TOGGLES_STATE = {
    'toggle slug': {
        'enabled': {'domain1', 'domain2},
        'disabled': {'domain4', 'domain3},
    }
}
```

sravfeyn commented:
@snopoke 

Re. making this customisable for other envs & domains, I did consider that, but practically I couldn't imagine any other env needing this in the near future. To the point of not over-engineering I decided to make this smallest change to get it working specifically for ICDS. I am happy to change it to as you suggest if you think that's necessary.

sravfeyn committed:
mutable default values

snopoke commented:
@sravfeyn I'm aware that no other projects may need this soon (though I'm sure the 3 other national scale projects that are getting going could use it). My desire to make it more general is not just so that other projects can use it but also to try and get away from custom ICDS changes creeping into the codebase. Perhaps it is simpler and better to just make this change as is to avoid over engineering but in general we should be moving to a more *pluggable* approach to customisations.

I'll approve this for now so that we can move ahead with other things but I'd like to hear what others think too @dannyroberts @czue @millerdev @ctsims

ctsims commented:
@snopoke I do think overall it's a bad look for us in Ethiopia and Burkina to have more built in rather than extensible behavior, but I think it depends a ton on which behaviors we are talking about.

A lot of the "ICDS" flags are more like "You are running CommCareHQ in single-tenant Enterprise mode" as much as they are "enable the CAS dashboard." In general it would be good to keep that in mind when we make changes like this to make sure we are separating things we turn on for CAS for scale reasons v. things we are turning on for CAS for custom project reasons.

millerdev commented:
I like @snopoke's suggestion of a generic datastructure in localsettings.py:
```py
STATIC_TOGGLES_STATE = {
    'toggle slug': {
        'enabled': {'domain1', 'domain2},
        'disabled': {'domain4', 'domain3},
    }
}
```
This does not seem like too much over-engineering. IMO the tiny complexity cost of making this feature generic is well worth it to avoid domain-specific code.

dannyroberts commented on corehq/toggles.py:
```diff
@@ -106,8 +106,12 @@ def __init__(self, slug, label, tag, namespaces=None, help_link=None,
+        self.always_enabled = (always_enabled or set() |
+            {'icds-cas'} if self.slug in settings.ICDS_CAS_ALWAYS_ENABLED_TOGGLES else {})
+        self.always_disabled = (always_disabled or set() |
+            {'icds-cas'} if self.slug in settings.ICDS_CAS_ALWAYS_DISABLED_TOGGLES else {})
```
nit: I think this should say `... else set()` so it's a `set` and not a `dict`. Not sure if the way it's called it makes a difference, but it could.

dannyroberts reviewed (commented):
I agree with what appears to be a consensus, that there's nothing inherently ICDS-specific going on here and we should set a simple convention that other projects can use to do the same thing, and that we should be generally attempting to trend to fewer specific mentions of the 'icds' env / 'icds-cas' project space in the general code base rather than more.

sravfeyn commented:
Thanks for all the thoughts. I do like making more general purpose things over custom, but rarely do I see the general things actually getting used in other envs/situations than the env/situation at the hand.

Anyway, this is classic dichotomy in writing code and I am more than happy to make this more general!

sravfeyn committed:
make it more general purpose

sravfeyn commented:
Updated. Also updated the relevant commcare-cloud code which gets a just a bit unclean to fit this format. https://github.com/dimagi/commcare-cloud/pull/3435/files

snopoke commented on corehq/toggles.py:
```diff
@@ -106,8 +106,12 @@ def __init__(self, slug, label, tag, namespaces=None, help_link=None,
+        # For icds-cas domain, toggles are delcared in localsettings statically
+        #   to avoid cache lookups
+        self.always_enabled = (always_enabled or set() |
+            settings.STATIC_TOGGLE_STATES.get(self.slug, {}).get('always_enabled', []))
```
```suggestion
            set(settings.STATIC_TOGGLE_STATES.get(self.slug, {}).get('always_enabled', [])))
```

snopoke commented on corehq/toggles.py:
```diff
@@ -106,8 +106,12 @@ def __init__(self, slug, label, tag, namespaces=None, help_link=None,
         self.save_fn = save_fn
-        self.always_enabled = always_enabled or set()
-        self.always_disabled = always_disabled or set()
+        # For icds-cas domain, toggles are delcared in localsettings statically
```
```suggestion
        # Toggles can be delcared in localsettings statically
```

snopoke commented on corehq/toggles.py:
```diff
@@ -106,8 +106,12 @@ def __init__(self, slug, label, tag, namespaces=None, help_link=None,
+        self.always_enabled = (always_enabled or set() |
+            settings.STATIC_TOGGLE_STATES.get(self.slug, {}).get('always_enabled', []))
+        self.always_disabled = (always_disabled or set() |
+            settings.STATIC_TOGGLE_STATES.get(self.slug, {}).get('always_disabled', []))
```
```suggestion
            set(settings.STATIC_TOGGLE_STATES.get(self.slug, {}).get('always_disabled', [])))
```

snopoke commented on settings.py:
```diff
@@ -27,6 +27,18 @@
+#   }
+STATIC_TOGGLE_STATES = {}
+
+STATIC_TOGGLE_STATES
```
```suggestion
```

sravfeyn replied:
Sorry I missed these after a context switch, thanks!

sravfeyn committed:
fix

sravfeyn commented:
Updated

millerdev commented on corehq/toggles.py:
```diff
@@ -106,8 +106,12 @@ def __init__(self, slug, label, tag, namespaces=None, help_link=None,
+        #   to avoid cache lookups
+        self.always_enabled = (always_enabled or set() |
+            set(settings.STATIC_TOGGLE_STATES.get(self.slug, {}).get('always_enabled', [])))
+        self.always_disabled = (always_disabled or set() |
```
nit: why combine with empty set?
```suggestion
        self.always_disabled = (always_disabled or
```
Also, `or` operator should be moved to the line below to avoid stickler comment.

The same applies two lines above.

dannyroberts replied:
I'm guessing the intention here was `(always_disabled or set()) | set(...))` and not `always_disabled or (set() | set(...))`. The code does the latter (I had to check in my python shell to be sure!), but was maybe intended to do the former.

sravfeyn replied:
Thanks for the catch, I was aware of this but I must have fumbled up the braces latter and missed it.

sravfeyn replied:
https://github.com/dimagi/commcare-hq/pull/26120

millerdev reviewed (commented):
The precedence of applying `always_(enabled|disabled)` passed into the constructor vs using `settings.STATIC_TOGGLE_STATES` is an interesting concern. One scenario I can think of is that someone working on SaaS would hard-code `always_enabled={'some-random-domain'}` in this file somewhere. With the current implementation the setting in code will override the value in `settings.STATIC_TOGGLE_STATES`, even when deployed to enterprise environments!

Now that we have a way to set these things using localsettings, maybe we should move all existing hard-coded values into localsettings on the relevant envs and remove the constructor parameters from `StaticToggle`? I think this is the best option and has least potential for confusion.

Other options:
- make `settings.STATIC_TOGGLE_STATES` take precedence over the constructor value.
- combine the constructor parameter and value from `settings.STATIC_TOGGLE_STATES` into a single set.

sravfeyn commented:
@millerdev Good point, please see https://github.com/dimagi/commcare-hq/pull/26120
